Converting the management cluster list and detail pages to be paginated

codyrancher opened the pull request:
<!-- This template is for Devs to give QA details before moving the issue To-Test -->
### Summary
Fixes #9548
<!-- Define findings related to the feature or bug issue. -->

### Occurred changes and/or fixed issues
Reduces the number of `findAll` calls and rely on the PaginatedResourceTable to do the paginated calls.

### Areas or cases that should be tested
The management cluster list and detail pages

### Areas which could experience regressions
The management cluster list and detail pages

### Checklist
- [ ] The PR is linked to an issue and the linked issue has a Milestone, or no issue is needed
- [ ] The PR has a Milestone <!-- The milestone should automatically be assigned if the linked issue has one, but does not always happen (incorrectly linked, issue has no milestone, etc) -->
- [ ] The PR template has been filled out
- [ ] The PR has been self reviewed <!-- There are no TODOs, no incorrect files in the PR, all the required files are there, no commented out code, etc-->
- [ ] The PR has a reviewer assigned
- [ ] The PR has automated tests or clear instructions for manual tests and the linked issue has appropriate QA labels, or tests are not needed
- [ ] The PR has reviewed with UX and tested in light and dark mode, or there are no UX changes

codyrancher commented on shell/detail/provisioning.cattle.io.cluster.vue:
```diff
@@ -283,97 +178,8 @@ export default {
       return info;
     },
 
-    fakeMachines() {
```
I haven't seen the behavior described in the comment. It appears were able to scale up and down and the machines/machine pools sync properly without all of these extra hoops.

richard-cox replied:
this is like the project/namespace list were we should show the group even if there are no rows. for instance if a pool has been scaled down to zero but we should still show the pool group. Does this still work with SSP off and on?

codyrancher commented on shell/detail/provisioning.cattle.io.cluster.vue:
```diff
@@ -243,15 +138,15 @@ export default {
   },
 
-  watch: {
-    showNodes(neu) {
```
I don't think we'll need this anymore especially with the new socket code.

codyrancher commented on shell/detail/provisioning.cattle.io.cluster.vue:
```diff
@@ -101,93 +78,12 @@ export default {
-      if ( this.$store.getters['management/canList'](MANAGEMENT.NODE_TEMPLATE) ) {
-        fetchTwo.nodeTemplates = this.$store.dispatch('management/findAll', { type: MANAGEMENT.NODE_TEMPLATE });
-      }
+      this.clusterToken = await this.value.getOrCreateToken();
```
I can prolly await on both this and the hasLocalCluster to speed things up slightly.

codyrancher commented on shell/detail/provisioning.cattle.io.cluster.vue:
```diff
@@ -29,42 +26,22 @@ import Socket, {
- * This class allows us to fake a machine deployment - when created, we set additional properties (_cluster etc)
- * and use these in the getters.
- **/
-class EmptyCapiMachineDeployment extends CapiMachineDeployment {
```
No longer used without the machinepool finagling.

codyrancher commented on shell/detail/provisioning.cattle.io.cluster.vue:
```diff
@@ -691,26 +543,6 @@ export default {
       }
     },
-
-    machineSortGenerationFn() {
```
This appears to only be necessary if we have the fake machines and nodes but neither appear to be necessary.

codyrancher commented on shell/detail/provisioning.cattle.io.cluster.vue:
```diff
@@ -745,33 +577,22 @@ export default {
     >
       <Tab
-        v-if="showMachines"
+        v-if="showMachines || true"
```
This will be removed. Accidentally left this in.

codyrancher commented on shell/list/provisioning.cattle.io.cluster.vue:
```diff
@@ -34,68 +33,29 @@ export default {
-      return filterHiddenLocalCluster(this.rows, this.$store);
+    // We no longer make the request for rows the paginated table does. We still need to access these rows for things like harvester.
+    filteredRows() {
+      return this.$refs.paginatedTable?.filteredRows || [];
```
Any thoughts on a better way to do this? I don't quite thing getters get us there unless we want to duplicate the filtering.

codyrancher replied:
Do we actually need to change anything for `hiddenHarvesterCount` and `tokenExpiredData`?  

Does harvester support SSP? Also, Is `tokenExpiredData`  Harvester specific? I don't see the banner label keys in our translations. Will SSP for Harvester be supported before the Harvester specific code is removed from shell?

I disabled it for `nonStandardNamespaces`.

richard-cox replied:
- hiddenHarvesterCount - this is used to help guide the user to manage their harvester cluster in the harvester plugin (as harvester clusters are not shown in the cluster managment cluster list). we need total of all applicable clusters - total of kube clusters (remaining are harvester clusters). Given `rows` no longer represents all clusters we would need determine both totals abother way
- tokenExpiredData - used to show a warning if any cluster has cluster creds associated which are expiring/expiring. we get this info from model data, so can't make an alternative request to fetch. we could either change it to warn given only the current page OR drop entirely for SSP world (would need to be a listed regression)
- nonStandardNamespaces - this can stay, but we would need to make a http request with pageSize =1 and filters for not fleet-local or fleet-default namespaces. there will be a total in the response. we could do this just once on fetch. alternatively though, and just realised, we can also just look at the COUNTS resource which will have a summary of resources per namespace

both hiddenHarvesterCount and tokenExpiredData cases are currently for harvester clusters (though the second uses a generic mechanism that can be expanded to other cluster types). requests to support them are handled by steve (even if they are harvester resources) so are supported by SSP.

codyrancher commented on shell/utils/pagination-utils.ts:
```diff
@@ -95,15 +95,7 @@ class PaginationUtils {
         return setting === enabledFor.resource?.id;
       }
 
-      if (setting.resource === enabledFor.resource?.id) {
-        if (!!setting.context) {
-          return enabledFor.resource?.context ? setting.context.includes(enabledFor.resource.context) : false;
```
It's not obvious to me why this is necessary? It seems to cause problems if you want to use the PaginatedResourceTable on non-list pages.

richard-cox replied:
This whole block needs to stay.

`if (setting.resource === enabledFor.resource?.id) {` - used to ensure the correct entry is found
`if (!!setting.context) {` - the context was used to explictely block pagination outside of supported areas. For instance the prov cluster was only allowed in the side nav and home page and not the cluster management list

codyrancher commented on shell/utils/settings.ts:
```diff
@@ -104,5 +104,11 @@ export const getPerformanceSetting = (rootGetters: Record<string, (arg0: string,
-  return Object.assign(safeDefaults, perfSetting || {});
+  const r = Object.assign(safeDefaults, perfSetting || {});
+
+  r.serverPagination.stores.cluster.resources.enableSome.enabled.push('provisioning.cattle.io.cluster');
+  r.serverPagination.stores.cluster.resources.enableSome.enabled.push('cluster');
+  r.serverPagination.stores.cluster.resources.enableSome.enabled.push('rke.cattle.io.etcdsnapshot');
```
Temporary to allow paginated requests

richard-cox replied:
in theory the three above shouldn't change anything as they're in the `cluster` store rather than `management`.

to get these without code changes, and fix the prov and mgmt cluster contexts (which are now not needed) `shell/config/settings.ts` `serverPagination` can be updated and then the new button `Populate with latest pagination defaults` can be used (and then applied). That process will soon be greatly updated (i have a PR, but it's in flux)

richard-cox commented on shell/detail/provisioning.cattle.io.cluster.vue:
```diff
@@ -574,6 +379,91 @@ export default {
+
+        if (this.$store.getters[`management/paginationEnabled`]?.(args)) {
+          const templateOpt = PaginationParamFilter.createMultipleFields([{
+            field: 'spec.clusterName',
```
this reference, and snapshot and machine deployments need adding to StevePaginationUtils.VALID_FIELDS and then also to https://github.com/rancher/rancher/issues/48603

codyrancher replied:
Done for both

richard-cox reviewed (changes requested):
It doesn't look like the list currently makes SSP requests. I fixed this by comment in shell/utils/settings.ts file.

Once enabled there were failed http calls to secondary resources given the filter wasn't supported by certain type (this is expected, added inline comment). i commented those filters out and the list shows however as there's no configured pagination the columns come from the schema attribute columns rather than matching what was there before. there's a fair few examples on how the pagination headers can be setup in the explorer product

codyrancher commented:
> 

It doesn't look like the list currently makes SSP requests. I fixed this by comment in shell/utils/settings.ts file.

Once enabled there were failed http calls to secondary resources given the filter wasn't supported by certain type (this is expected, added inline comment). i commented those filters out and the list shows however as there's no configured pagination the columns come from the schema attribute columns rather than matching what was there before. there's a fair few examples on how the pagination headers can be setup in the explorer product

Yeah, I really don't know what happened there. Anyway, it should be working now.

codyrancher commented:
> This whole block needs to stay.
> if (setting.resource === enabledFor.resource?.id) { - used to ensure the correct entry is found
if (!!setting.context) { - the context was used to explictely block pagination outside of supported areas. For instance the prov cluster was only allowed in the side nav and home page and not the cluster management list

I put this back in. I think it might be nice if we add dev warnings so it's a little more obvious why some resources aren't using the pagination API

richard-cox commented on shell/plugins/dashboard-store/actions.js:
```diff
@@ -143,6 +143,21 @@ export default {
     }
   },
 
+  async findPageOrAll(ctx, {
```
Interesting point. I've just used findPage when fetching secondary resources.

We can switch to this approach though

richard-cox replied:
Lets go down this path.

I've just hit a similar scenario with findMatching. if pagination is on for type it would use findPage instead.

i was thinking of flipping the order, so we would having

- findAll
- findAllOrPage
- findMatching
- findMatchingOrPage

codyrancher replied:
I mean... if we're going into the paginated world developers will have to understand the difference. Is there a way to get around that?

If we wanted to slow roll this and get devs slowly accustomed I think we've already gone down the wrong path. To slow roll I would've done the feature flag switches on a page level so we could keep the exact same logic on the existing pages and then switch to the paginated pages if the feature flag is set. In this scenario we would've had a two way door where we could turn the feature flag off (maybe a per page global setting) and fallen back on existing tested logic. Since we're switching on the component level we now have paginated and unpaginated code mixed throughout so I think the devs will have to recognize and understand it.

One other option instead of the `findAllorPage` method could just be dependency inversion with an interface like `find(notPaginatedFindMethod, paginatedFindMethod)`. In this scenario the consumer has to understand what method's they're invoking and we're just doing the feature flag switching for them. It's a little clunky with the store though.

richard-cox commented on shell/plugins/dashboard-store/actions.js:
```diff
@@ -143,6 +143,21 @@ export default {
   },
 
+  async findPageOrAll(ctx, {
+    type, context, paginatedOpt, findAllOpt
```
not that findAllOpts isn't supplied where this is used.

however we should be safe just chopping out the pagination object and using the paginatedOpts (though it should be harmless)

codyrancher replied:
I'm not sure I'm understanding the ask here. 

Are you just wanting me to remove the object de-structuring?

richard-cox commented on shell/utils/cluster.js:
```diff
@@ -63,6 +63,26 @@ export function paginationFilterHiddenLocalCluster(store, filterMgmtCluster = tr
   return PaginationParamFilter.createMultipleFields(filter);
 }
 
+export async function hasAccessToLocalCluster(store) {
```
this could just use `management/find`

codyrancher replied:
Switched

richard-cox replied:
not sure, but i would have hoped fetching an individual resource was highly optimised, even if it does have an extra hop to kube & etcd.

TopLevelMenu has been converted to use SSP (there's some pending stuff from backend though). findPage was intended to only be used in SSP world, mainly for a websocket update reason but soon for others

richard-cox commented on shell/list/provisioning.cattle.io.cluster.vue:
```diff
@@ -34,68 +35,30 @@ export default {
-    this.mgmtClusters = res.mgmtClusters;
-  },
-
   data() {
```
there's an error in console

```
[Vue warn]: Property "loadResources" was accessed during render but is not defined on instance.
```
Looks like loadResources comes from the resource list mixins, however even before these changes i dont think it was exposed?

codyrancher replied:
it was exposed by the resourceFetch mixin. It's no longer necessary since it's the PaginatedResourceTable that's doing the fetching.

I've removed it.

richard-cox reviewed (changes requested):
There's an error in console

```
Error in fetch(): TypeError: Cannot read properties of null (reading 'fields')
    at steve-pagination-utils.ts:433:1
```

one of the filter values is `null`. there'll be a bug somewhere, probably somewhere generic. let me know if you want me to track it down.

With latest `head` i'm seeing an error when a filter is used (`"unrecognized operator: ",`). Will check with the BE team

codyrancher commented on shell/list/provisioning.cattle.io.cluster.vue:
```diff
@@ -179,12 +362,6 @@ export default {
 
 <template>
   <div>
-    <Banner
```
Put it back in

codyrancher commented:
> There's an error in console

Error in fetch(): TypeError: Cannot read properties of null (reading 'fields')
    at steve-pagination-utils.ts:433:1

one of the filter values is null. there'll be a bug somewhere, probably somewhere generic. let me know if you want me to track it down.

This one turned out to be my fault. I missed it because I had largely stopped verifying  the `hide-local-cluster ` setting and it was only surfacing when that was on.

codyrancher commented on shell/config/settings.ts:
```diff
@@ -278,8 +278,10 @@ export const DEFAULT_PERF_SETTING: PerfSettings = {
           enableSome: {
             enabled: [
-              { resource: CAPI.RANCHER_CLUSTER, context: ['home', 'side-bar'] },
+              { resource: CAPI.RANCHER_CLUSTER, context: ['home', 'side-bar', 'provisioning.cattle.io.clusters'] },
```
I removed it but I know I needed it for my current version of rancher. I haven't tried the latest since we had that sort issue you pointed out.

richard-cox replied:
interesting. there's the filter issue in latest rancher. lets take a look at this once that's fixed (waiting on steve bump, though delayed a bit because i asked for https://github.com/rancher/rancher/issues/48603 at same time)

richard-cox replied:
regarding the latest update, i made a mistake. 

leaving only the prov cluster context in will mean the side nav won't use pagination. which means there'll be no error given the filtering that the side nav uses is (backend) broken.

so the final version of this line would need to have no context

codyrancher commented on shell/config/settings.ts:
```diff
@@ -278,8 +278,10 @@ export const DEFAULT_PERF_SETTING: PerfSettings = {
-              { resource: CAPI.RANCHER_CLUSTER, context: ['home', 'side-bar'] },
+              { resource: CAPI.RANCHER_CLUSTER, context: ['home', 'side-bar', 'provisioning.cattle.io.clusters'] },
               { resource: MANAGEMENT.CLUSTER, context: ['side-bar'] },
+              { resource: CAPI.MACHINE, context: ['provisioning.cattle.io.clusters', 'provisioning.cattle.io.cluster'] },
```
I removed it but I know I needed it for my current version of rancher. I haven't tried the latest since we had that sort issue you pointed out.

codyrancher commented on shell/list/provisioning.cattle.io.cluster.vue:
```diff
@@ -167,7 +130,227 @@ export default {
-    }
+    },
+
+    headers() {
```
I updated all of the paginated headers. We seem to use formatters and getters pretty extensively so I largely just disabled search and sort.

Are the fields we can sort/search by documented somewhere? Or will that be the same fields that are in `steve-pagination-utils`? I would assume not since we likely want to do different indexing on the fields depending on the use case.

richard-cox replied:
Looks good. Remember to track the removal or sort/filter on these filters in the PR description (i'll sum them up in a single document). To be honest these are one's i think we'll need an alternative solution for (something from BE), but not right now.

Fields that can be sorted/filtered by aka indexed fields are indeed those ones in steve-pagination-utils. I'll update the comment to include a bit of extra info

```
  /**
   * Filtering/Sorting with the vai cache can only be executed on specific fields
   * 1) Those listed here
   * 2) Those references in the schema's attributes.fields list (which is used by generic lists)
   *
   * This list will track rancher/steve pkg/stores/sqlproxy/proxy_store.go and also any pending entries yet to be indexed. It's used to provide dev time validation of sort/filter fields
   */
  static VALID_FIELDS: { [type: string]: { field: string, startsWith?: boolean }[]} = {
```

codyrancher commented on shell/components/form/ResourceTabs/index.vue:
```diff
@@ -85,13 +85,13 @@ export default {
+        eventSchema ? headerFromSchemaColString('Subobject', eventSchema, this.$store.getters, true) : null,
+        eventSchema ? headerFromSchemaColString('Source', eventSchema, this.$store.getters, true) : null,
         MESSAGE,
-        headerFromSchemaColString('First Seen', eventSchema, this.$store.getters, true),
-        headerFromSchemaColString('Count', eventSchema, this.$store.getters, true),
+        eventSchema ? headerFromSchemaColString('First Seen', eventSchema, this.$store.getters, true) : null,
+        eventSchema ? headerFromSchemaColString('Count', eventSchema, this.$store.getters, true) : null,
         STEVE_NAME_COL,
-      ]
+      ].filter((header) => header)
```
Apparently users may not have access to the events. I noticed this when testing against a user that didn't have access to the local cluster. Let me know if you'd like to resolve this in another way.

codyrancher commented:
@richard-cox except for the findPaginatedOrAll conversation there aren't anymore changes you want here are there? This is just backend blocked.

I just want to make sure there's nothing else that needs to be actioned upon before I move on.

richard-cox commented on shell/utils/cluster.js:
```diff
@@ -63,6 +63,24 @@ export function paginationFilterHiddenLocalCluster(store, filterMgmtCluster = tr
 }
 
+/**
+ * Checks to see if the current user has access to the local store
```
minor typo - `local store` --> `local cluster`

richard-cox commented on shell/detail/provisioning.cattle.io.cluster.vue:
```diff
@@ -435,7 +248,40 @@ export default {
+          search: false
+        },
+        {
+          ...ROLES,
```
this comes from a label, so we should be able to sort/filter via ssp

richard-cox replied:
(same for mgmt nodes)

richard-cox commented on shell/detail/provisioning.cattle.io.cluster.vue:
```diff
@@ -435,7 +248,40 @@ export default {
+          search: false
+        },
+        {
+          ...MACHINE_NODE_OS,
```
we might be able to sort on this, though it would be a bit off with the empty value (which defaults to linux)

richard-cox replied:
(same for mgmt nodes)

richard-cox commented on shell/detail/provisioning.cattle.io.cluster.vue:
```diff
@@ -520,6 +385,30 @@ export default {
+      return [
+        {
+          ...STEVE_STATE_COL,
+          value:         'snapshotFile.status',
```
is this indexed? if not we need to add it

richard-cox commented on shell/detail/provisioning.cattle.io.cluster.vue:
```diff
@@ -1055,6 +1054,40 @@ export default {
+          context="provisioning.cattle.io.cluster"
+        >
+          <!-- Removing the group by namespace/list buttons in the middle header -->
+          <template #header-middle />
```
We specify a groupBy (backup location), i think that the controls should still show (given it's an annotation that can be indexed)

richard-cox reviewed (changes requested):
I managed to miss the detail page, given it a once over code wise. 

Functional wise we're still blocked on the BE filter bug

nwmac commented:
@richard-cox @codyrancher  Is this still needed? I have pushed to 2.13, but please update and close if stale. Thanks
